Add alembic functionality to plugins

djsilcock opened the pull request:
This adds alembic functionality to the plugins and adds three `click` commands

`flaskbb plugins migrations <pluginname>` - creates migration files in plugins/<plugin_dir>/migration_versions

`flaskbb plugins upgrade <pluginname>` - upgrades db with plugin models (ie installs) from the generated migrations
`flaskbb plugins downgrade <pluginname>` - downgrades db plugin models (ie uninstalls)

Also adds a db converter to allow the migration routine to find `Table` definitions (classes deriving from `Model` are found automatically)

see example_model_plugin for example usage

sh4nks commented:
This is really nice, thanks! I hope I can take a closer look tomorrow (and on your other PR as well)!

sh4nks commented on flaskbb/cli/plugins.py:
```diff
@@ -128,3 +128,29 @@ def list_plugins():
+@plugins.command("migrations")
+@click.argument("plugin_identifier")
+def migrate_plugin(plugin_identifier):
+    """Installs a new plugin."""
```
"Perform database migrations for a plugin" or something a long the lines would be better

sh4nks commented on flaskbb/cli/plugins.py:
```diff
@@ -128,3 +128,29 @@ def list_plugins():
+@plugins.command("upgrade")
+@click.argument("plugin_identifier")
+def upgrade_plugin(plugin_identifier):
+    """Uninstalls a plugin from FlaskBB."""
```
"Upgrade the plugin to the latest version" or something similar is more appropriate

sh4nks commented on flaskbb/plugins/__init__.py:
```diff
@@ -10,8 +10,35 @@
+import os,copy
+
+@contextlib.contextmanager
+def plugin_name_migrate(name):
```
could you fix the pep8 erros and add some comments/docstrings in the next 3 methods (plugin_name_migrate, db_for_plugin, config_migrate)? My editor automatically highlights the pep8 errors.

djsilcock commented:
Have made some changes here
The actual migration files are already in the individual plugin folders - plugins/migrations only actually contains the environment information. The plugins use the named-branch functionality of alembic >0.7 so all the migrations files themselves are in the migrations_versions folder created by `flaskbb plugins migrations` within the individual plugin folder. Plugins can still be dragged and dropped into another flaskbb installation without copying any files to the main migrations folder. I have renamed the folder to make it less ambiguous.
I have deleted the example plugin completely - the docstrings explain the syntax. A `db=db_for_plugins(__name__,extensions.db` probably belongs in the cookiecutter file - I will PR it there.
Some PEP8 changes have been made and docstrings updated

sh4nks commented:
Awesome and thanks for the little recap!
Did I understand this correctly, that each plugin will use a different branch so that they won't conflict with each other? I didn't know that something like this is possible with alembic.
Let me know when this is ready to merge.

djsilcock commented:
It uses the multiple bases feature described here: 
http://alembic.zzzcomputing.com/en/latest/branches.html#running-with-multiple-bases

sh4nks commented:
> Tests and proper docs still need to be added

@djsilcock any updates on this? Otherwise I'll slowly start preparing the first release.

djsilcock commented:
Sorry - have been busy with other things - should have it done and tested in a couple of days

djsilcock commented:
not sure why the test_new_plugin test is not working - the test migrate test fails because it doesn't like being run twice (the temporary files are getting persisted by py.test. They both work on my own (Windows) system

sh4nks commented:
Thanks! I'll have a closer look (and test it) after the weekend!

sh4nks commented on flaskbb/plugins/__init__.py:
```diff
@@ -32,15 +121,24 @@ def installable(self):
 
+    def this_version_installed(self):
+        if self.uninstallable():
+            if Setting.as_dict(self.settings_key).get('version',None)==self.version:
```
This indicates that we expect to store the current version in the database as well? Wouldn't it better if we compare the installed alembic version? And it looks like that this isn't used anywhere.

sh4nks commented on flaskbb/plugins/__init__.py:
```diff
@@ -8,18 +8,107 @@
 
 
-class FlaskBBPlugin(Plugin):
+def db_for_plugin(plugin_name,sqla_instance=None):
```
What about this function? I haven't seen it in use anywhere.

sh4nks commented on flaskbb/plugins/__init__.py:
```diff
@@ -8,18 +8,107 @@
+            fname='"%s"'%fname
+        return fname
+
+    def get_migration_version_dir(self):
```
Same for this. Instead you have always used ``os.path.join(self.path,'migration_versions')``

sh4nks commented:
@djsilcock, I finally had some more time today to have a closer look at your PR! I have added some comments to some methods where I am not sure what the purpose of those is - because they are never used (I was also looking at the tests to see if they are used there)

It would also be really great if you could add a little example (a simple GitHub Gist would be enough which I could incorporate into the docs) on how to use those new features.

sh4nks commented:
I did some clean ups (especially PEP8 fixes, and seriously man, you gotta make some spaces ;)) on your branch merged it. I will still have to do a few things like making the tests not fail, adding some docs how to use it and some other cleanups. This is (soon, was) actually the last issue which kept me from releasing a first 1.0 version.
